### DISCUSSION TITLE
golang no longer building on linux

### DISCUSSION THREAD

**Discussion by tundratim** [Homebrew/discussions#336] [Linux] [answered]
When attempting to install go to a non-default location in linux, I see this error:

`fatal: unable to access 'https://go.googlesource.com/tools.git/': error setting certificate verify locations:  CAfile: /home/linuxbrew/.linuxbrew/etc/openssl@1.1/cert.pem CApath: /home/linuxbrew/.linuxbrew/etc/openssl@1.1/certs
`

I can work around this with:

`git config --global http.sslVerify false
`
However, this was never necessary in the past.  It feels like the git build logic has the default cert location wired into it. 

UPDATE:  This is specifically and only a problem with https based clones.  ssh clones seems to work fine.  So, it's the https certs that are causing the noise.

In this case, git was built from source with brew.  If I uninstall it and use system git (in this case ubuntu) there is no problem - system git is finding whatever cert authority info it needs.  So ... I'd surmise that this has something do do with how git gets built on linux.

**Comment by tundratim** [comment]
The more I poke at this, the more this looks like an openssl@1.1 problem, but I have no clue how to fix

**Reply by gromgit** [reply]
What does `brew info openssl@1.1` output?

**Reply by tundratim** [reply]
`[test] dock4  ~>brew info openssl@1.1
Error: Please update your system cURL.
Minimum required version: 7.41.0
Your cURL version: 7.29.0
Your cURL executable: /usr/bin/curl
openssl@1.1: stable 1.1.1i
Cryptography and SSL/TLS Toolkit
https://openssl.org/
/opt/TundraWare/tools/Cellar/openssl@1.1/1.1.1i (8,403 files, 24.7MB) *
  Built from source on 2020-12-21 at 00:13:19
/opt/TundraWare/tools/Library/Homebrew/shims/scm/git --version
From: https://github.com/Homebrew/linuxbrew-core/blob/HEAD/Formula/openssl@1.1.rb
License: OpenSSL
==> Caveats
A CA file has been bootstrapped using certificates from the system
keychain. To add additional certificates, place .pem files in
  /opt/TundraWare/tools/etc/openssl@1.1/certs

and run
  /opt/TundraWare/tools/opt/openssl@1.1/bin/c_rehash
/usr/bin/curl --disable --globoff --show-error --user-agent Linuxbrew/2.6.2-214-g2ddb71a\ \(Linux\;\ x86_64\ 5.4.0-58-generic\)\ curl/7.29.0 --header Accept-Language:\ en --retry 3 --max-time 5 https://formulae.brew.sh/api/formula-linux/openssl@1.1.json
==> Analytics
==> install (30 days)
Index | Name (with options)                                                                                          |  Count |  Percent
-----:|--------------------------------------------------------------------------------------------------------------|-------:|--------:
1     | openssl@1.1                                                                                                  | 23,819 |  100.00%
==> install (90 days)
Index | Name (with options)                                                                                          |  Count |  Percent
-----:|--------------------------------------------------------------------------------------------------------------|-------:|--------:
1     | openssl@1.1                                                                                                  | 75,485 |  100.00%
==> install (365 days)
Index | Name (with options)                                                                                         |   Count |  Percent
-----:|-------------------------------------------------------------------------------------------------------------|--------:|--------:
1     | openssl@1.1                                                                                                 | 292,683 |  100.00%
==> install-on-request (30 days)
Index | Name (with options)                                                                                           | Count |  Percent
-----:|---------------------------------------------------------------------------------------------------------------|------:|--------:
1     | openssl@1.1                                                                                                   |   647 |  100.00%
==> install-on-request (90 days)
Index | Name (with options)                                                                                           | Count |  Percent
-----:|---------------------------------------------------------------------------------------------------------------|------:|--------:
1     | openssl@1.1                                                                                                   | 1,772 |  100.00%
==> install-on-request (365 days)
Index | Name (with options)                                                                                           | Count |  Percent
-----:|---------------------------------------------------------------------------------------------------------------|------:|--------:
1     | openssl@1.1                                                                                                   | 6,121 |  100.00%
==> build-error (30 days)
Index | Name (with options)                                                                                           | Count |  Percent
-----:|---------------------------------------------------------------------------------------------------------------|------:|--------:
1     | openssl@1.1   
`

**Reply by gromgit** [reply]
And what do `brew config` and `brew doctor` output?

**Comment by jonchang** [comment]
My guess is that `brew postinstall openssl@1.1` failed, but you didn't notice the error. Try running that.

**Reply by tundratim** [reply]
Per @danielnachun brew reinstall -s curl seemed to resolve the issue for the moment, but I am not sure why that works.

**Comment by danielnachun** [comment]
I'm fairly sure this error is happening because `curl` is currently set to `cellar :any`, but it is not relocatable without some configuration by the user: https://github.com/Linuxbrew/legacy-linuxbrew/issues/380#issuecomment-190344899.  For now I think `brew reinstall -s curl` will fix your problem, but we should figure out how to deal with the cellar status of `curl` to avoid this type of issue.

**Reply by tundratim** [reply]
This seems to fix the problem, but I am not sure why.  What is -s doing that a normal source install doesn't?

**Reply by gromgit** [reply]
> What is -s doing that a normal source install doesn't?

Nothing, but if you didn't use `-s` in your curl install, it's quite likely that it was _not_ from source (that's what @danielnachun is hinting at with `cellar :any`).

**Reply by tundratim** [reply]
Hmmm, I was fairly sure that everything I installed (via makefile) got built from source.  

Anyway, I will update the make file to use -s for all installs to force source-based installations.

**Reply by carlocab** [reply]
If it just does `brew install <formula>` then that doesn't always build from source. If you're putting it in a script such as a makefile, you may wish to use the longer flag `--build-from-source` for clarity.

**Reply by gromgit** [reply]
> Anyway, I will update the make file to use -s for all installs to force source-based installations.

That would be a good idea, since you're running Homebrew on Linux with a non-standard `HOMEBREW_PREFIX`. Homebrew actually tries very hard _not_ to build from source, and a `cellar :any` tag in a formula basically says "you can install me from bottle in any directory"...even when there are hardcoded paths hidden in the binaries.

**Reply by tundratim** [reply]
So upon further digging, I found the mechanism of failure.  Even if you do brew install -s curl,  it immediately goes about installing git as one of its dependencies.  But git itself lists curl as a dependency and when IT install curl, it does not do so from source.  So, merely adding -s to all install commands won't work.  Sequencing is required:

brew -s install git
brew -s reinstall curl

Is there some sort of ENVVAR I could set to force brew to always install everything from source so that even dependent packages get properly built?

**Reply by carlocab** [reply]
Not at the moment, ~but this should be fixed in a recent PR. Let try to track it down~ no. (I misremembered regarding the PR I was thinking about.)

**Reply by danielnachun** [reply]
In my experience on a non-standard `HOMEBREW_PREFIX` on Linux I've not had any issues with relocatable bottles except this recent issue with `curl`.  If you try to build a `curl` bottle locally it will tell you it is not relocatable, and also says the same in CI.  I don't know why `curl` currently has `cellar :any` but that only showed up with 7.74.  Up until then, `curl` was always built from source with a non-standard `HOMEBREW_PREFIX`.  I think this recent change may been an accidental oversight.  I already have a [PR](https://github.com/Homebrew/linuxbrew-core/pull/21873) open for `curl` for an unrelated issue and will bring this up there.

**Reply by tundratim** [reply]
I've also found that after just pulling the brew repo and doing `brew install gcc`  that a ton of stuff is suddenly getting built. For some reason, brew is not seeing that system curl and git are on its path and proceeds to  build curl with the problems noted above.  It's been a while since I built this project from  scratch but I do not recall a simple gcc install - even at a nonstandard location invoking all this extra machinery.

**Reply by danielnachun** [reply]
It's probably doing that because your system `curl` is too old (see the warning in the output of `brew info` you posted above).

**Comment by tundratim** [comment, marked as answer]
RESOLVED: While the process above did expose curl not being properly installed from source,  the underlying problem that exposed  this also had to be addressed.  The docker container being used for builds was using CentOS7.  As @danielnachun  pointed out, the versions of git and curl there were so far backleveled, brew kept trying to install its own which is when all the serial dependencies and failure to build from source consistently showed up.

The fix was to use a craftier approach to get around the broken curl and all is now well.
